<a href="https://colab.research.google.com/github/ChatMadhu/Sonar-rock-vs-Mine-prediction/blob/main/Project_1_SONAR_Rock_vs_Mine_Prediction_with_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import the libraries

In [ ]:
import numpy as np                                    #this is for creating numpy arrays
import pandas as pd                                   #this is for data loading into pandas data frame
from sklearn.model_selection import train_test_split  #this is to devide the data set into two as training and testing data sets
from sklearn.linear_model import LogisticRegression   #this is to select the algorithem, which is used in this project
from sklearn.metrics import accuracy_score            #this is to find the accuracy of the created model

Data collection and data processing

In [ ]:
sonar_data = pd.read_csv('/content/sonar data.csv', header = None)    #first we have to load our csv data set into pandas dataframe.
                                                                      #If you check you can find that there is no headers in the dcsv data set. so it have to be mentioned for this pandas data frame also.


In [ ]:
sonar_data.head()         #shows the first 5 rows

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [ ]:
#lets find how much data rows are there

sonar_data.shape

(208, 61)

In [ ]:
sonar_data.describe()                             #this describes the statistical measures of data.

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [ ]:
sonar_data[60].value_counts()           #here the 60 means that we have to specify that the 60th colum is the one we're dealing here.

M    111
R     97
Name: 60, dtype: int64

M=> Mine
R=> Rock
#since the 111 and 97 almost equal ane we will get a good prediction and also a good accuracy
#more the data more accurate the model is


In [ ]:
sonar_data.groupby(60).mean()     #by this we're going to find the mean value for each row and using that value we are going to predict the rock or minne

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [ ]:
# Lets seperate the data and lables. This is a supervised learning design. So we're labeling the values
#it means the data and the target have to be seperated. The data will be labeled as 0 to 207 and the target is to be labeled as M and R.

X = sonar_data.drop(columns=60,  axis=1)
Y = sonar_data[60]
print(X)
print(Y)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

Lets seperate the data set as training and testing data. 

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.1, stratify = Y, random_state = 1)

# here the X and Y is the training and testing data. And 0.1 test size is that we are using 10% of the data set for the training. the stratify means that we need to split the data according to  the y values. it  means for the testing data and also for trainig data, the y parameters should be spreaded evenly.
#random state means the splitting style of data by the algorithem. if we split the data as random_state=1, it splits data in one way. For random state=2 it splits the data in another style and give another portion of data as training and also for testing.

In [ ]:
print(X.shape, X_test.shape, X_train.shape)
print(X_train)
print(Y_train)

(208, 60) (21, 60) (187, 60)
         0       1       2       3       4       5       6       7       8   \
115  0.0414  0.0436  0.0447  0.0844  0.0419  0.1215  0.2002  0.1516  0.0818   
38   0.0123  0.0022  0.0196  0.0206  0.0180  0.0492  0.0033  0.0398  0.0791   
56   0.0152  0.0102  0.0113  0.0263  0.0097  0.0391  0.0857  0.0915  0.0949   
123  0.0270  0.0163  0.0341  0.0247  0.0822  0.1256  0.1323  0.1584  0.2017   
18   0.0270  0.0092  0.0145  0.0278  0.0412  0.0757  0.1026  0.1138  0.0794   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
140  0.0412  0.1135  0.0518  0.0232  0.0646  0.1124  0.1787  0.2407  0.2682   
5    0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   
154  0.0117  0.0069  0.0279  0.0583  0.0915  0.1267  0.1577  0.1927  0.2361   
131  0.1150  0.1163  0.0866  0.0358  0.0232  0.1267  0.2417  0.2661  0.4346   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   

         9   ...      

Model Training ------> Logistic Regression

In [ ]:
#now lets train the data with logistic regression model

model = LogisticRegression()
model.fit(X_train, Y_train)

LogisticRegression()

In [ ]:
# now the model should be evaluated

# 1) for that first we should find the accuracy on testing data. If the accuracy is more than 70% it is good

X_train_prediction = model.predict(X_train)                           #X train prediction is the prediction of our model.
training_data_accuracy = accuracy_score(X_train_prediction, Y_train) #Y train is the actual answer of the predicting data. from this accuracy score, the predicted values and the actual values are compared each and give an accuracy for that results.
print(training_data_accuracy, '= Training data Accuracy')


0.8342245989304813 = Training data Accuracy


In [ ]:
# now we will see how much is the testing data accuracy.
X_test_prediction = model.predict(X_test)                           #X train prediction is the prediction of our model.
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test) #Y train is the actual answer of the predicting data. from this accuracy score, the predicted values and the actual values are compared each and give an accuracy for that results.
print(testing_data_accuracy, '= Testing data Accuracy')

0.7619047619047619 = Testing data Accuracy


Making a predictive system

In [ ]:
# first we should input some values

input_data = (0.0084,0.0153,0.0291,0.0432,0.0951,0.0752,0.0414,0.0259,0.0692,0.1753,0.1970,0.1167,0.1683,0.0814,0.2179,0.5121,0.7231,0.7776,0.6222,0.3501,0.3733,0.2622,0.3776,0.7361,0.8673,0.8223,0.7772,0.7862,0.5652,0.3635,0.3534,0.3865,0.3370,0.1693,0.2627,0.3195,0.1388,0.1048,0.1681,0.1910,0.1174,0.0933,0.0856,0.0951,0.0986,0.0956,0.0426,0.0407,0.0106,0.0179,0.0056,0.0236,0.0114,0.0136,0.0117,0.0060,0.0058,0.0031,0.0072,0.0045
)

# now we have to convert the data array in to a  numpy array

input_data_as_numpy_array = np.asarray(input_data)


#now we have to reshape the numpyh array as we are predicting for one instance

input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = model.predict(input_data_reshaped)

print(prediction)

['R']


This R is included in a list. So we can modify this result as if the prediction = R, then the modelo is 

In [ ]:
if(prediction[0]== 'R'):
  print('The objetc us a ROck')
else:
  print('The object is a Mine') 

The objetc us a ROck
